media/media-source/media-managedmse-memorypressure-inactive.html is flaky.

jyavenard opened the pull request:
#### 174634df51d32a92ba9e7af449f5283338f705ab
<pre>
media/media-source/media-managedmse-memorypressure-inactive.html is flaky.
<a href="https://bugs.webkit.org/show_bug.cgi?id=253560">https://bugs.webkit.org/show_bug.cgi?id=253560</a>
rdar://106416033

Reviewed by Jer Noble.

The test exposed an underlying problem related to how we notified the
SourceBuffer that its buffered attribute was &quot;dirty&quot;.
When the SourceBufferPrivate modified its buffered range, it would signal
the SourceBufferPrivateClient (here the SourceBuffer) that the buffered
range became dirty and needed to be recalculated.
However, in WK2 and if the GPU process is enabled, the SourceBufferClient
is a RemoteSourceBufferProxy: the new buffered range
wasn&apos;t actually sent back to the content process, only that it became dirty.
If we attempted to read the SourceBuffer&apos;s buffered attribute (operation
happening in the content process) then we could return the stale old value.
This would be particularly problematic during the Prepare Append algorithm
(<a href="https://w3c.github.io/media-source/#sourcebuffer-prepare-append)">https://w3c.github.io/media-source/#sourcebuffer-prepare-append)</a> where
eviction could occur: the buffered attribute if read then would be the
wrong value and not indicating that an eviction occurred.
As this code is excercised via MockMediaSource which never runs in the
GPU Process, this issue wasn&apos;t identified earlier (will be handled in webkit.org/b/253884).

Notifying the SourceBufferPrivate became dirty was an unnecessary operation
- There are limited cases during which the source buffer&apos;s buffered attribute can
  change:
    - during the prepare append algorithm
    - after an append operation
    - after a remove operation
    - if ManagedSourceBuffer, after a memory pressure signal.

We only have to update the buffered range at the completion of those
operations, and ensure that we always return an up to date value.
MediaSource.buffered is modified to only calculate the HTMLMediaElement
buffered attribute when it is read.

* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::isBuffered const):
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::buffered):
(WebCore::MediaSource::onReadyStateChange):
(WebCore::MediaSource::updateBufferedIfNeeded):
(WebCore::MediaSource::buffered const): Deleted.
(WebCore::MediaSource::sourceBufferDidChangeBufferedDirty): Deleted.
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::sourceBufferPrivateBufferedChanged):
(WebCore::SourceBuffer::sourceBufferPrivateBufferedDirtyChanged): Deleted.
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/platform/graphics/MediaSourcePrivate.h:
(WebCore::MediaSourcePrivate::bufferedChanged): Deleted.
* Source/WebCore/platform/graphics/MediaSourcePrivateClient.h:
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::setBufferedRanges): Use move semantics.
(WebCore::SourceBufferPrivate::updateBufferedFromTrackBuffers):
(WebCore::SourceBufferPrivate::clearTrackBuffers):
(WebCore::SourceBufferPrivate::removeCodedFrames):
(WebCore::SourceBufferPrivate::didReceiveSample):
(WebCore::SourceBufferPrivate::memoryPressure):
(WebCore::SourceBufferPrivate::setBufferedDirty): Deleted.
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
* Source/WebCore/platform/graphics/SourceBufferPrivateClient.h:
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:
(WebKit::RemoteMediaSourceProxy::buffered):
(WebKit::RemoteMediaSourceProxy::buffered const): Deleted.
(WebKit::RemoteMediaSourceProxy::bufferedChanged): Deleted.
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.messages.in:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::evictCodedFrames): Return updated buffered range.
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateBufferedDirtyChanged): Deleted.
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.messages.in: Return updated buffered range.
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
(WebKit::MediaSourcePrivateRemote::bufferedChanged): Deleted.
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::evictCodedFrames):
(WebKit::SourceBufferPrivateRemote::updateBufferedFromTrackBuffers):
(WebKit::SourceBufferPrivateRemote::removeCodedFrames):
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateAppendComplete):
(WebKit::SourceBufferPrivateRemote::memoryPressure):
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateBufferedDirtyChanged): Deleted.
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.messages.in:

Canonical link: <a href="https://commits.webkit.org/261801@main">https://commits.webkit.org/261801@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/99507e6bf2d016b430c44d1b66def76f8a5c213a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112910 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22069 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/90/builds/1586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/4687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/121408 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/117012 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23416 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/13230 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/87/builds/4687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118697 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17389 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/90/builds/1586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/106000 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/99349 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/90/builds/1586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/46412 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/14364 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/90/builds/1586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15068 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/13230 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/20380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/90/builds/1586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8233 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16919 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
```diff
@@ -88,9 +88,10 @@ bool ManagedMediaSource::isBuffered(const PlatformTimeRanges& ranges) const
         return true;
 
     ASSERT(ranges.length() == 1);
+    ASSERT(m_buffered);
```
This should just be an early `return false;`

jyavenard replied:
ok

jernoble commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -100,10 +94,13 @@ void SourceBufferPrivate::updateHighestPresentationTimestamp()
 {
-    m_buffered->ranges() = timeRanges;
-    setBufferedDirty(true);
+    if (m_buffered->ranges() == timeRanges)
```
This is a pre-existing issue, but we should change `RefPtr<PlatformTimeRanges> m_buffered` -> `Ref<PlatformTimeRanges>`.

jyavenard replied:
done

jyavenard replied:
actually, I'll make the buffered() method return a reference instead and have the member be a Ref<T>. but it's a much bigger change. Will do so in a follow up change

jyavenard committed:
media/media-source/media-managedmse-memorypressure-inactive.html is flaky.
https://bugs.webkit.org/show_bug.cgi?id=253560
rdar://106416033

Reviewed by Jer Noble.

The test exposed an underlying problem related to how we notified the
SourceBuffer that its buffered attribute was "dirty".
When the SourceBufferPrivate modified its buffered range, it would signal
the SourceBufferPrivateClient (here the SourceBuffer) that the buffered
range became dirty and needed to be recalculated.
However, in WK2 and if the GPU process is enabled, the SourceBufferClient
is a RemoteSourceBufferProxy: the new buffered range
wasn't actually sent back to the content process, only that it became dirty.
If we attempted to read the SourceBuffer's buffered attribute (operation
happening in the content process) then we could return the stale old value.
This would be particularly problematic during the Prepare Append algorithm
(https://w3c.github.io/media-source/#sourcebuffer-prepare-append) where
eviction could occur: the buffered attribute if read then would be the
wrong value and not indicating that an eviction occurred.
As this code is excercised via MockMediaSource which never runs in the
GPU Process, this issue wasn't identified earlier (will be handled in webkit.org/b/253884).

Notifying the SourceBufferPrivate became dirty was an unnecessary operation
- There are limited cases during which the source buffer's buffered attribute can
  change:
    - during the prepare append algorithm
    - after an append operation
    - after a remove operation
    - if ManagedSourceBuffer, after a memory pressure signal.

We only have to update the buffered range at the completion of those
operations, and ensure that we always return an up to date value.
MediaSource.buffered is modified to only calculate the HTMLMediaElement
buffered attribute when it is read.

* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::isBuffered const):
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::buffered):
(WebCore::MediaSource::onReadyStateChange):
(WebCore::MediaSource::updateBufferedIfNeeded):
(WebCore::MediaSource::buffered const): Deleted.
(WebCore::MediaSource::sourceBufferDidChangeBufferedDirty): Deleted.
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::sourceBufferPrivateBufferedChanged):
(WebCore::SourceBuffer::sourceBufferPrivateBufferedDirtyChanged): Deleted.
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/platform/graphics/MediaSourcePrivate.h:
(WebCore::MediaSourcePrivate::bufferedChanged): Deleted.
* Source/WebCore/platform/graphics/MediaSourcePrivateClient.h:
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::setBufferedRanges): Use move semantics.
(WebCore::SourceBufferPrivate::updateBufferedFromTrackBuffers):
(WebCore::SourceBufferPrivate::clearTrackBuffers):
(WebCore::SourceBufferPrivate::removeCodedFrames):
(WebCore::SourceBufferPrivate::didReceiveSample):
(WebCore::SourceBufferPrivate::memoryPressure):
(WebCore::SourceBufferPrivate::setBufferedDirty): Deleted.
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
* Source/WebCore/platform/graphics/SourceBufferPrivateClient.h:
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:
(WebKit::RemoteMediaSourceProxy::buffered):
(WebKit::RemoteMediaSourceProxy::buffered const): Deleted.
(WebKit::RemoteMediaSourceProxy::bufferedChanged): Deleted.
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.messages.in:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::evictCodedFrames): Return updated buffered range.
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateBufferedDirtyChanged): Deleted.
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.messages.in: Return updated buffered range.
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
(WebKit::MediaSourcePrivateRemote::bufferedChanged): Deleted.
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::evictCodedFrames):
(WebKit::SourceBufferPrivateRemote::updateBufferedFromTrackBuffers):
(WebKit::SourceBufferPrivateRemote::removeCodedFrames):
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateAppendComplete):
(WebKit::SourceBufferPrivateRemote::memoryPressure):
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateBufferedDirtyChanged): Deleted.
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.messages.in:

Canonical link: https://commits.webkit.org/261801@main
